### PR TITLE
Added/Updated Approve(), SetResult(), Vote() and Finalize() funcs to UI.

### PR DESCRIPTION
**Pull request by chrisli30**

### CODE REVIEW

File: src/components/bodhi-dls/cardFinished.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    const {
+      amount, config, children, result,
+    } = this.props;
+
+    const amountStr = amount.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
+
+    // Build Alert elements based on result
+    let alertElement;
+
+    if (result) {
```
Comment (dwalintukan) [inline src/components/bodhi-dls/cardFinished.js]: is  `result` ever true here? don't see any logic to set `result` before this?

File: src/containers/oracle.js [hunk 2]
```diff
@@ -66,124 +90,128 @@ class OraclePage extends React.Component {
...
-
+    // contractAddress should be the address of this oracle
+    // TODO: Update this to oracle.address for testing on testnet
     const contractAddress = '9697b1f2701ca9434132723ee790d1cb0ab0e414';
-    const selectedIndex = oracle.optionIdxs[radioValue - 1];
-    const senderAddress = 'qKjn4fStBaAtwGiwueJf9qFxgpbAvf1xAy';
-    console.log(`contractAddress is ${contractAddress}, selectedIndex is ${selectedIndex}, senderAddress is ${senderAddress}`);
 
-    this.props.onSetResult(contractAddress, selectedIndex, senderAddress);
+    if (this.state.config.name === 'BETTING') {
```
Comment (dwalintukan) [inline src/containers/oracle.js]: change to switch block

File: src/containers/oracle.js [hunk 3]
```diff
@@ -66,124 +90,128 @@ class OraclePage extends React.Component {
...
+    } else if (this.state.config.name === 'VOTING') {
+      /** The amount of voting needs to be approved by Bodhi_token * */
+      onApprove(oracle.topicAddress, amount, senderAddress);
+
+      setTimeout(() => {
+        onVote(contractAddress, selectedIndex, amount, senderAddress);
+      }, SUB_REQ_DELAY);
+
+      // The last Vote that hits BOT threshold needs to issue a finalize() after certain delay
+      if (oracle.amounts[selectedIndex] + amount >= ORACLE_BOT_THRESHOLD) {
```
Comment (dwalintukan) [inline src/containers/oracle.js]: I am confused by this logic. Why is it calling `finalizeResult()` if it passes the threshold? Is this for the demo?

Normally, only finalizeResult() is called after the DecentralizedOracle does not hit the threshold within the `arbitrationEndBlock` period. This sets the final result in `TopicEvent`.

File: src/containers/topic.js [hunk 4]
```diff
@@ -91,15 +82,22 @@ class TopicPage extends React.Component {
...
-      value: `${topic.amounts[index]} ${topic.token}`,
-      percent: _.floor((topic.amounts[index] / totalBalance) * 100),
+    const qtumTotal = _.sum(topic.qtumAmount);
+    const qtumBalance = _.map(topic.qtumAmount, (amount, idx) => ({
+      name: topic.options[idx],
+      value: `${amount} QTUM`,
+      percent: _.floor((amount / qtumTotal) * 100),
+    }));
+
+
```
Comment (dwalintukan) [inline src/containers/topic.js]: 👾

File: src/redux/topic/actions.js [hunk 5]
```diff
@@ -16,6 +22,18 @@ const topicActions = {
...
   }),
+
+  VOTE: 'VOTE',
+  VOTE_RETURN: 'VOTE_RETURN',
+  onVote: (contractAddress, index, amount, senderAddress) => ({
+    type: topicActions.VOTE,
+    payload: {
+      contractAddress,
+      index,
+      amount,
```
Comment (dwalintukan) [inline src/redux/topic/actions.js]: `botAmount` to match the API param

File: src/redux/topic/saga.js [hunk 6]
```diff
@@ -74,6 +103,42 @@ export function* setResultRequestHandler() {
   });
 }
 
+export function* voteRequestHandler() {
+  yield takeEvery(actions.VOTE, function* onVoteRequest(action) {
+    const {
+      contractAddress,
+      index,
+      amount,
```
Comment (dwalintukan) [inline src/redux/topic/saga.js]: botAmount to match API param

File: src/redux/topic/saga.js [hunk 7]
```diff
@@ -74,6 +103,42 @@ export function* setResultRequestHandler() {
...
+      senderAddress,
+    } = action.payload;
+
+    try {
+      const options = {
+        method: 'POST',
+        body: JSON.stringify({
+          contractAddress,
+          index,
+          amount,
```
Comment (dwalintukan) [inline src/redux/topic/saga.js]: botAmount

**Comment by dwalintukan** [comment]
The timer based system for `approve()` > `setResult()` and `approve()` > `vote()` will be fine for the demo. But we need to figure out better logic for the MVP release.

We can't assume a block will be mined (with the approve) in x number of seconds as it can change in the future. So we need to do actual validation on txids before we `setResult()` and `vote()`.

My suggestion (for setResult example):
1. User clicks on set result button > executes `approve()` in API
2. Save `txid` from approve() call
3. UI button is grayed out and disabled. Change text to "Approving 100 BOT... please wait"
4. We start a polling service, say every 30 secs, hit an API endpoint that validates if the `txid` has been mined. And that it was accepted.
5. Once the txid is validated, we automatically call `setResult()`
6. UI button is grayed out and disabled. Change text to "Setting result... please wait"
7. Start another polling service to validate `setResult` txid
8. Once valid, hide the action button permanently

File: src/components/bodhi-dls/cardFinished.js (hunk 1 above)
Comment (chrisli30) [reply src/components/bodhi-dls/cardFinished.js]: Yeah, my bad. Made an update to mark it required and passed down from topic.js

File: src/containers/oracle.js (hunk 3 above)
Comment (chrisli30) [reply src/containers/oracle.js]: Updated with new logic which checks if current block is greater than arbitrationEndBlock

File: src/containers/topic.js [hunk 8]
```diff
@@ -91,15 +82,22 @@ class TopicPage extends React.Component {
...
+    const qtumTotal = _.sum(topic.qtumAmount);
+    const qtumBalance = _.map(topic.qtumAmount, (amount, idx) => ({
+      name: topic.options[idx],
+      value: `${amount} QTUM`,
+      percent: qtumTotal === 0 ? qtumTotal : _.floor((amount / qtumTotal) * 100),
+    }));
+
+
+    const botTotal = _.sum(topic.botAmount);
+    const botBalance = _.map(topic.botAmount, (amount, idx) => ({
```
[the remark this answers is not in the document]
Comment (chrisli30) [reply src/containers/topic.js]: Not yet, but will probably add another progress bar. I'll leave it here for now.

File: src/redux/app/reducer.js [hunk 9]
```diff
@@ -36,6 +36,11 @@ export default function appReducer(state = initState, action) {
 
       return state.set('walletAddrs', result);
     }
+
+    /** Block Count * */
+    case actions.GET_BLOCK_COUNT_RETURN:
+      return state.set('get_block_count_return', action.value);
```
[the remark this answers is not in the document]
Comment (chrisli30) [reply src/redux/app/reducer.js]: Thought about your comment, but I want to make all state variables use underlines, but differentiate them with normal variables. Will make them all in same style in subsequent PR.

File: src/containers/oracle.js [hunk 10]
```diff
@@ -66,124 +96,154 @@ class OraclePage extends React.Component {
...
+
+      case 'FINALIZING':
+        // Finalize oracle to enter next stage, withdraw
+        onFinalizeResult(contractAddress, senderAddress);
+        break;
 
-    this.props.onSetResult(contractAddress, selectedIndex, senderAddress);
+      default:
+
+        // TODO: oracle not found page
```
[the remark this answers is not in the document]
Comment (chrisli30) [reply src/containers/oracle.js]: Correct, if oracle not found elements will not render on this page. This line will not run. Removed.

**Pull request merged**